### PR TITLE
feat(coupon): added coupon fraud use case

### PR DESCRIPTION
**Pull request by Orkuncakilkaya**

### CODE REVIEW

File: pages/coupon-fraud/index.js [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    useEffect(() => {
+        !isWaitingForReponse && messageRef.current?.scrollIntoView({behavior: 'smooth'});
+    }, [isWaitingForReponse]);
+
+    async function handleSubmit(e) {
+        e.preventDefault();
+        setIsWaitingForReponse(true);
+
+        const fpQuery = await visitorDataQuery.refetch();
+        console.log({status: fpQuery.status, error: fpQuery.error, fpQuery});
```
Comment (TheUnderScorer) [inline pages/coupon-fraud/index.js:43]: Remember to remove this `console.log` before we deploy it to production.

File: pages/coupon-fraud/index.js [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        }
+        setAuthMessage(responseJson.message);
+        setSeverity(responseJson.severity);
+        setHttpResponseStatus(responseStatus);
+        setIsWaitingForReponse(false);
+    }
+
+    return (
+        <UseCaseWrapper
+            sx={{width: '100%'}}
```
Comment (TheUnderScorer) [inline pages/coupon-fraud/index.js:75]: This style is unnecessary, this component has styles by default with `width: 100%`

File: pages/coupon-fraud/index.js [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+        const responseJson = await response.json();
+        const responseStatus = response.status;
+        if (responseStatus === 200) {
+            setPrice(1000);
+        }
+        setAuthMessage(responseJson.message);
+        setSeverity(responseJson.severity);
+        setHttpResponseStatus(responseStatus);
+        setIsWaitingForReponse(false);
```
Comment (TheUnderScorer) [inline pages/coupon-fraud/index.js:70]: In most recent use cases, we've been using the `useMutation` hook, which makes the code more readable. You can check it for example here: `client/api/loan-risk/use-request-loan.js`

File: pages/coupon-fraud/coupon-fraud.module.css [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    flex-direction: row;
+    justify-content: space-between;
+}
+
+.productInfo {
+    display: flex;
+    flex-direction: column;
+}
+
+.productTitle {
```
Comment (TheUnderScorer) [inline pages/coupon-fraud/coupon-fraud.module.css:22]: Let's remove empty selectors

File: pages/coupon-fraud/index.js [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+  const [couponCode, setCouponCode] = useState();
+  const [price, setPrice] = useState(1119);
+
+  const handleSubmit = useCallback(
+    async (event) => {
+      event.preventDefault();
+      const fpData = await visitorDataQuery.refetch();
+
+      console.log({ fpData: fpData.data });
```
Comment (TheUnderScorer) [inline pages/coupon-fraud/index.js:29]: Let's remove this `console.log`

**Review by TheUnderScorer — COMMENTED** [review]
Looks good, thanks for the fixes @Orkuncakilkaya! I left only a couple of minor comments.

Let's keep this PR until @makma returns from his PTO so he can review the texts.

File: pages/index.js [hunk 6]
```diff
@@ -46,6 +46,14 @@ As a result, you will protect your users and your business against various payme
     
     Demonstration of personalized content such as search history, customized user interface, or even shopping cart.`,
   },
+  {
```
Comment (makma) [inline pages/index.js:49]: Alphabetical order pls

File: pages/coupon-fraud/index.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    if (couponClaimMutation?.data?.severity === 'success') {
+      setPrice(1000);
+    }
+  }, [couponClaimMutation]);
+
+  const isLoading = visitorDataQuery.isLoading || couponClaimMutation.isLoading;
+
+  return (
+    <UseCaseWrapper
+      title="Coupon fraud problem"
```
Comment (makma) [inline pages/coupon-fraud/index.js:47]:
```suggestion
      title="Coupon Fraud problem"
```

File: pages/coupon-fraud/index.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+      description={<>This page demonstrates how to solve coupon fraud problem.</>}
+      listItems={[
+        <>You can apply coupon to this item only once</>,
```
Comment (makma) [inline pages/coupon-fraud/index.js:50]:
```suggestion
        <>You can apply a coupon to the specific item only once.</>,
```

File: pages/coupon-fraud/index.js [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        <>You can&apos;t apply same coupon more than once</>,
```
Comment (makma) [inline pages/coupon-fraud/index.js:51]:
```suggestion
```

File: pages/coupon-fraud/index.js [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        <>You can&apos;t spam coupon codes, there is a 1 hour threshold</>,
```
Comment (makma) [inline pages/coupon-fraud/index.js:52]:
```suggestion
        <>You cannot spam coupon codes, there is a 1 hour threshold.</>,
```

File: pages/coupon-fraud/index.js [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        <>You can&apos;t apply the same coupon in incognito mode</>,
```
Comment (makma) [inline pages/coupon-fraud/index.js:53]:
```suggestion
        <>You cannot apply the same coupon in incognito mode.</>,
```

File: pages/coupon-fraud/index.js [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+        <>
+          Sample coupon codes are <code>123456</code> and <code>098765</code>
```
Comment (makma) [inline pages/coupon-fraud/index.js:55]:
```suggestion
          Sample coupon codes are <code>123456</code> and <code>098765</code>.
```

File: pages/api/coupon-fraud/claim.js [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+async function getVisitorClaim(visitorId, couponCode) {
+  return await CouponClaim.findOne({
+    where: { visitorId, couponCode },
+  });
+}
+
+/**
+ * Checks if given visitor has already used a coupon code.
```
Comment (makma) [inline pages/api/coupon-fraud/claim.js:27]: This comment (or function) is not correct. There's no visitorId in the function.

File: pages/api/coupon-fraud/claim.js [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+        [Op.eq]: code,
+      },
+    },
+  });
+}
+
+/**
+ * Saves coupon code into database. If it already exists, we update its timestamp.
+ */
+export async function saveCouponCode(visitorId, couponCode) {
```
Comment (makma) [inline pages/api/coupon-fraud/claim.js:42]: This function does not save or store a coupon. It applies a coupon (or creates a claim). I'd rename it and also improved the comment accordingly.

File: pages/api/coupon-fraud/claim.js [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+
+export default couponEndpoint(async (req, res, { visitorId, couponCode }) => {
+  const coupon = await checkCoupon(couponCode);
+
+  // Check if the coupon exists
+  if (!coupon) {
+    return getForbiddenReponse(res, 'Coupon code not exists', 'error');
+  }
+
+  const claimed = await getVisitorClaim(visitorId, couponCode);
```
Comment (makma) [inline pages/api/coupon-fraud/claim.js:61]: I know it's really subjective but I'd name this const `wasCouponClaimedByVisitor` or similar.

File: pages/api/coupon-fraud/claim.js [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+
+  // Check if the visitor claimed this coupon before
+  if (claimed) {
+    return getForbiddenReponse(res, 'Visitor used this coupon before', 'error');
+  }
+
+  const recentClaim = await checkVisitorClaimedRecently(visitorId);
```
Comment (makma) [inline pages/api/coupon-fraud/claim.js:68]: It's really confusing and hard to understand the difference between `claimed` and `recentClaim`. Try to remove, be verbose if necessary.

File: pages/api/coupon-fraud/claim.js [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+
+  // Check if the visitor claimed a coupon recently
+  if (recentClaim) {
+    return getForbiddenReponse(res, 'Visitor claimed another coupon recently', 'error');
+  }
+
+  await saveCouponCode(visitorId, couponCode);
```
Comment (makma) [inline pages/api/coupon-fraud/claim.js:75]: Does it save coupon code or rather apply a claim? I believe the later, consider renaming.

File: pages/coupon-fraud/index.js [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+          <div className={styles.productContainer}>
+            <div className={styles.productItem}>
+              <div className={styles.productInfo}>
+                <Typography fontSize={20}>iPhone 14 Pro Max 256 GB - Deep Purple</Typography>
+                <div className={styles.productImage}>
+                  <Image width={256} height={256} src="/iphone14.png" alt="iPhone image" />
+                </div>
+              </div>
+              <Typography fontWeight="bold">
+                {new Intl.NumberFormat('en-US', { currency: 'USD', style: 'currency' }).format(price)}
```
Comment (makma) [inline pages/coupon-fraud/index.js:70]: Nice touch with `new Intl.NumberFormat...`!

File: pages/coupon-fraud/index.js [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+              </Typography>
+            </div>
+          </div>
+          <div className={styles.couponContainer}>
+            <Typography>Do you have a coupon? Apply to get discount!</Typography>
```
Comment (makma) [inline pages/coupon-fraud/index.js:75]:
```suggestion
            <Typography>Do you have a coupon? Apply to get a discount!</Typography>
```

**Review by makma — CHANGES_REQUESTED** [review]

File: pages/coupon-fraud/index.js (hunk 18 above)
Comment (Orkuncakilkaya) [reply pages/coupon-fraud/index.js:69]: Thank you!

File: pages/api/coupon-fraud/claim.js [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+    where: {
+      code: {
+        [Op.eq]: code,
+      },
+    },
+  });
+}
+
+/**
+ * Claim coupon behalf of visitor
```
Comment (makma) [inline pages/api/coupon-fraud/claim.js:40]: Please check all the comments you created with some grammar checker, e.g. Grammarly. There are a lot of missing articles and prepositions.

```suggestion
 * Claim coupon on behalf of the visitor.
```

File: pages/api/coupon-fraud/claim.js [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+
+  return claim;
+}
+
+export default couponEndpoint(async (req, res, { visitorId, couponCode }) => {
+  const coupon = await checkCoupon(couponCode);
+
+  // Check if the coupon exists
+  if (!coupon) {
+    return getForbiddenReponse(res, 'Coupon code not exists', 'error');
```
Comment (makma) [inline pages/api/coupon-fraud/claim.js:58]: Check and fix grammar in all strings
```suggestion
    return getForbiddenReponse(res, 'Provided coupon code does not exist.', 'error');
```

File: pages/api/coupon-fraud/claim.js [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+  const wasCouponClaimedByVisitor = await getVisitorClaim(visitorId, couponCode);
+
+  // Check if the visitor claimed this coupon before
+  if (wasCouponClaimedByVisitor) {
+    return getForbiddenReponse(res, 'Visitor used this coupon before', 'error');
+  }
+
+  const visitorClaimedAnotherCouponRecently = await checkVisitorClaimedRecently(visitorId);
+
+  // Check if the visitor claimed a coupon recently
```
Comment (makma) [inline pages/api/coupon-fraud/claim.js:70]: Since you've correctly renamed this the const, this comment is not needed anymore.

**Review by makma — CHANGES_REQUESTED** [review]
Description in the tile is weak, doesn't look good IMO.
Change it to: 

> While increased sales are good for your business, customers acting in bad-faith can dry up available promotional campaign dollars by repeatedly using coupon codes.

> Preventing customers, either signed-in or guests, from abusing your coupons and promotions can help you prevent financial losses and increase sales.

File: pages/api/coupon-fraud/claim.js (hunk 22 above)
Comment (Orkuncakilkaya) [reply pages/api/coupon-fraud/claim.js]: Removed 👍

**Review by makma — APPROVED** [review]

**Pull request merged**